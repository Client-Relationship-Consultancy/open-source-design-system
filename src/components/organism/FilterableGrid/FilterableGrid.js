import React from "react"
import styled from "styled-components"
import PropTypes from "prop-types"
import { colourPalette } from "brandColours"

const SearchContainer = styled.div`
  margin-bottom: 1.5rem;
  display: flex;
  align-items: center;
`
SearchContainer.displayName = "SearchContainer"

const Label = styled.label`
  color: ${props => props.theme.primary.main.hex};
`
Label.defaultProps = {
  theme: colourPalette.examplePalette,
}
Label.displayName = "Label"

const Search = styled.input`
  margin-left: 1rem;
  border: 1px solid ${props => props.theme.secondary.main.hex};
  padding: 0.5rem 1rem;
  border-radius: 5px;
  font-size: 1rem;
  color: ${props => props.theme.black.main.hex};
  ::placeholder {
    color: ${props => props.theme.black.tint80.hex};
  }
`
Search.defaultProps = {
  theme: colourPalette.examplePalette,
}
Search.displayName = "Search"

const GridContainer = styled.div`
  display: grid;
  grid-template-columns: ${props => `repeat(${props.itemsPerRow}, 1fr)`};
  grid-gap: 1rem;
  height: ${props => (props.height ? props.height : "auto")};
  overflow: auto;
`
GridContainer.displayName = "GridContainer"

// Though not used, it makes testing easier
const GridItem = styled.div``
GridItem.displayName = "GridItem"

export const NoResults = styled.span`
  color: ${props => props.theme.black.tint80.hex};
`
NoResults.defaultProps = {
  theme: colourPalette.examplePalette,
}
NoResults.displayName = "NoResults"

class FilterableGrid extends React.Component {
  state = {
    initialItems: this.props.items,
    items: this.props.items,
  }

  static getDerivedStateFromProps = (nextProps, prevState) => {
    if (nextProps.items !== prevState.initialItems) {
      return {
        initialItems: nextProps.items,
        items: nextProps.items,
      }
    }
    return null
  }

  filterItems = event => {
    const value = event.target.value.toLowerCase()
    this.setState(prevState => ({
      items: prevState.initialItems.filter(item => item.key.toLowerCase().includes(value)),
    }))
  }

  renderItems = items => {
    if (items && items.length > 0) {
      return items.map(item => <GridItem key={item.key}>{item.value}</GridItem>)
    }
    return <NoResults>{this.props.emptyMessage}</NoResults>
  }

  render() {
    const { label, itemsPerRow, height, id, className } = this.props
    return (
      <div className={className}>
        <SearchContainer>
          <Label>
            {label}
            <Search id={id} placeholder="Search" type="text" onChange={this.filterItems} />
          </Label>
        </SearchContainer>
        <GridContainer itemsPerRow={itemsPerRow} height={height}>
          {this.renderItems(this.state.items)}
        </GridContainer>
      </div>
    )
  }
}

FilterableGrid.displayName = "FilterableGrid"

FilterableGrid.propTypes = {
  itemsPerRow: PropTypes.number,
  items: PropTypes.array,
  label: PropTypes.string,
  height: PropTypes.string,
  emptyMessage: PropTypes.string,
  id: PropTypes.string,
  className: PropTypes.string,
}

FilterableGrid.defaultProps = {
  itemsPerRow: 5,
  label: "Search",
  emptyMessage: "No results",
}

export default FilterableGrid
