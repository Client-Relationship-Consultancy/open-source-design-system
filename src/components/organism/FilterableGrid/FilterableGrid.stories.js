import React from "react"
import { storiesOf } from "@storybook/react"
import { FilterableGrid } from "../../../index"

const stories = storiesOf("FilterableGrid", module)

const items = Array.from({ length: 40 }, (x, i) => ({
  key: `item ${i + 1}`,
  value: <div>Item {i + 1} </div>,
}))

const FilterableGridWrapper = () => (
  <div>
    <FilterableGrid itemsPerRow={6} items={items} height="150px" />
  </div>
)

stories.add("Default", context => {
  const text = {
    description: "This is the filterable grid component",
    dos: [
      "Use this component when there is a large list that needs to be filtered",
      "Specify how many on each row with the 'itemsPerRow' props",
      "You can input JSX elements into the 'items' array prop",
    ],
    donts: ["Don't squash all the items together. Put additional padding on each item if needed."],
  }
  Object.assign(context, text)
  return <FilterableGridWrapper />
})
