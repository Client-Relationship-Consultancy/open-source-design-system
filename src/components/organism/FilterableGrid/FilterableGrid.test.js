import React from "react"
import { shallow, mount } from "enzyme"
import "jest-styled-components"
import FilterableGrid, { NoResults } from "./FilterableGrid"

const labelText = "Filter"
const empty = "Empty"
const height = "150px"
const items = [
  { key: "foo", value: "foo" },
  { key: "bar", value: "bar" },
  { key: "hello", value: "hello" },
  { key: "world", value: "world" },
  { key: "minion", value: "minion" },
]
const newItems = [
  { key: "new", value: "new" },
  { key: "items", value: "items" },
]

describe("FilterableGrid component", () => {
  const component = shallow(<FilterableGrid label={labelText} emptyMessage={empty} items={items} />)
  it("render the correct number of items", () => {
    expect(component.find("GridContainer").children().length).toEqual(items.length)
    expect(component.find("GridItem").length).toEqual(items.length)
  })

  it("renders the correct messages", () => {
    const emptyComponent = shallow(<FilterableGrid label={labelText} emptyMessage={empty} />)
    expect(emptyComponent.find("Label").text()).toEqual(labelText)
    expect(emptyComponent.find("GridItem").length).toEqual(0)
    expect(emptyComponent.find("NoResults").text()).toEqual(empty)
  })

  it("renders the correct dynamic css", () => {
    const styledComponent = mount(
      <FilterableGrid itemsPerRow={4} label={labelText} emptyMessage={empty} items={items} />,
    )
    expect(styledComponent.find("GridContainer")).toHaveStyleRule(
      "grid-template-columns",
      "repeat(4,1fr)",
    )
    expect(styledComponent.find("GridContainer")).toHaveStyleRule("height", "auto")

    styledComponent.setProps({ height })
    expect(styledComponent.find("GridContainer")).toHaveStyleRule("height", height)
  })
  it("renders the correct items when searching", () => {
    component.find("Search").simulate("change", {
      target: {
        value: "foo",
      },
    })
    expect(component.find("GridItem").length).toEqual(1)
    expect(component.find("GridItem").text()).toEqual("foo")
  })
  it("renders no items when searching regex type expressions", () => {
    component.find("Search").simulate("change", {
      target: {
        value: "*",
      },
    })
    expect(component.find("GridItem").length).toEqual(0)
  })
  it("Rerender the props and it updates the state", () => {
    component.setProps({ items: newItems })
    expect(component.state().initialItems).toEqual(newItems)
  })
  it("should render no results component with correct styling", () => {
    const noResult = mount(<NoResults />)
    expect(noResult).toMatchSnapshot()
  })
})
