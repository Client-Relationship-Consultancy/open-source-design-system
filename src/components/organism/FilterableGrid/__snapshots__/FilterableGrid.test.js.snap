// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterableGrid component should render no results component with correct styling 1`] = `
.c0 {
  color: #696969;
}

<NoResults
  theme={
    Object {
      "action": Object {
        "dark": Object {
          "RGB": "79,38,98",
          "hex": "#4E2662",
          "on": "#FFFFFF",
        },
        "light": Object {
          "RGB": "173,101,207",
          "hex": "#AD65CF",
          "on": "#FFFFFF",
        },
        "main": Object {
          "RGB": "116,60,143",
          "hex": "#743C8F",
          "on": "#FFFFFF",
        },
      },
      "background": Object {
        "RGB": "246,246,246",
        "hex": "#F6F6F6",
        "on": "#272727",
      },
      "black": Object {
        "main": Object {
          "RGB": "39,39,39",
          "hex": "#272727",
          "on": "#FFFFFF",
        },
        "tint20": Object {
          "RGB": "236,236,236",
          "hex": "#ECECEC",
          "on": "#272727",
        },
        "tint40": Object {
          "RGB": "199,199,199",
          "hex": "#C7C7C7",
          "on": "#272727",
        },
        "tint60": Object {
          "RGB": "160,160,160",
          "hex": "#A0A0A0",
          "on": "#FFFFFF",
        },
        "tint80": Object {
          "RGB": "105,105,105",
          "hex": "#696969",
          "on": "#FFFFFF",
        },
      },
      "complimentary": Object {
        "dark": Object {
          "RGB": "206,185,0",
          "hex": "#CEB900",
          "on": "#272727",
        },
        "light": Object {
          "RGB": "255,245,157",
          "hex": "#FFF59D",
          "on": "#272727",
        },
        "main": Object {
          "RGB": "255,238,82",
          "hex": "#FFEE52",
          "on": "#272727",
        },
      },
      "error": Object {
        "dark": Object {
          "RGB": "197,73,1",
          "hex": "#C54901",
          "on": "#FFFFFF",
        },
        "main": Object {
          "RGB": "255,115,32",
          "hex": "#FF7320",
          "on": "#FFFFFF",
        },
      },
      "primary": Object {
        "dark": Object {
          "RGB": "102,22,20",
          "hex": "#661614",
          "on": "#FFFFFF",
        },
        "light": Object {
          "RGB": "218,52,47",
          "hex": "#DA342F",
          "on": "#FFFFFF",
        },
        "main": Object {
          "RGB": "164,40,36",
          "hex": "#A42824",
          "on": "#FFFFFF",
        },
      },
      "secondary": Object {
        "dark": Object {
          "RGB": "5,38,59",
          "hex": "#05263B",
          "on": "#FFFFFF",
        },
        "light": Object {
          "RGB": "89,156,175",
          "hex": "#599CAF",
          "on": "#FFFFFF",
        },
        "main": Object {
          "RGB": "11,70,107",
          "hex": "#0B466B",
          "on": "#FFFFFF",
        },
      },
      "surface": Object {
        "RGB": "255,255,255",
        "hex": "#FFFFFF",
        "on": "#272727",
      },
      "white": Object {
        "RGB": "255,255,255",
        "hex": "#FFFFFF",
        "on": "#272727",
      },
    }
  }
>
  <StyledComponent
    forwardedComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "_foldedDefaultProps": Object {
          "theme": Object {
            "action": Object {
              "dark": Object {
                "RGB": "79,38,98",
                "hex": "#4E2662",
                "on": "#FFFFFF",
              },
              "light": Object {
                "RGB": "173,101,207",
                "hex": "#AD65CF",
                "on": "#FFFFFF",
              },
              "main": Object {
                "RGB": "116,60,143",
                "hex": "#743C8F",
                "on": "#FFFFFF",
              },
            },
            "background": Object {
              "RGB": "246,246,246",
              "hex": "#F6F6F6",
              "on": "#272727",
            },
            "black": Object {
              "main": Object {
                "RGB": "39,39,39",
                "hex": "#272727",
                "on": "#FFFFFF",
              },
              "tint20": Object {
                "RGB": "236,236,236",
                "hex": "#ECECEC",
                "on": "#272727",
              },
              "tint40": Object {
                "RGB": "199,199,199",
                "hex": "#C7C7C7",
                "on": "#272727",
              },
              "tint60": Object {
                "RGB": "160,160,160",
                "hex": "#A0A0A0",
                "on": "#FFFFFF",
              },
              "tint80": Object {
                "RGB": "105,105,105",
                "hex": "#696969",
                "on": "#FFFFFF",
              },
            },
            "complimentary": Object {
              "dark": Object {
                "RGB": "206,185,0",
                "hex": "#CEB900",
                "on": "#272727",
              },
              "light": Object {
                "RGB": "255,245,157",
                "hex": "#FFF59D",
                "on": "#272727",
              },
              "main": Object {
                "RGB": "255,238,82",
                "hex": "#FFEE52",
                "on": "#272727",
              },
            },
            "error": Object {
              "dark": Object {
                "RGB": "197,73,1",
                "hex": "#C54901",
                "on": "#FFFFFF",
              },
              "main": Object {
                "RGB": "255,115,32",
                "hex": "#FF7320",
                "on": "#FFFFFF",
              },
            },
            "primary": Object {
              "dark": Object {
                "RGB": "102,22,20",
                "hex": "#661614",
                "on": "#FFFFFF",
              },
              "light": Object {
                "RGB": "218,52,47",
                "hex": "#DA342F",
                "on": "#FFFFFF",
              },
              "main": Object {
                "RGB": "164,40,36",
                "hex": "#A42824",
                "on": "#FFFFFF",
              },
            },
            "secondary": Object {
              "dark": Object {
                "RGB": "5,38,59",
                "hex": "#05263B",
                "on": "#FFFFFF",
              },
              "light": Object {
                "RGB": "89,156,175",
                "hex": "#599CAF",
                "on": "#FFFFFF",
              },
              "main": Object {
                "RGB": "11,70,107",
                "hex": "#0B466B",
                "on": "#FFFFFF",
              },
            },
            "surface": Object {
              "RGB": "255,255,255",
              "hex": "#FFFFFF",
              "on": "#272727",
            },
            "white": Object {
              "RGB": "255,255,255",
              "hex": "#FFFFFF",
              "on": "#272727",
            },
          },
        },
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "FilterableGrid__NoResults-sc-121qbq7-5",
          "isStatic": false,
          "lastClassName": "c0",
          "rules": Array [
            "color:",
            [Function],
            ";",
          ],
        },
        "displayName": "NoResults",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "FilterableGrid__NoResults-sc-121qbq7-5",
        "target": "span",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
    theme={
      Object {
        "action": Object {
          "dark": Object {
            "RGB": "79,38,98",
            "hex": "#4E2662",
            "on": "#FFFFFF",
          },
          "light": Object {
            "RGB": "173,101,207",
            "hex": "#AD65CF",
            "on": "#FFFFFF",
          },
          "main": Object {
            "RGB": "116,60,143",
            "hex": "#743C8F",
            "on": "#FFFFFF",
          },
        },
        "background": Object {
          "RGB": "246,246,246",
          "hex": "#F6F6F6",
          "on": "#272727",
        },
        "black": Object {
          "main": Object {
            "RGB": "39,39,39",
            "hex": "#272727",
            "on": "#FFFFFF",
          },
          "tint20": Object {
            "RGB": "236,236,236",
            "hex": "#ECECEC",
            "on": "#272727",
          },
          "tint40": Object {
            "RGB": "199,199,199",
            "hex": "#C7C7C7",
            "on": "#272727",
          },
          "tint60": Object {
            "RGB": "160,160,160",
            "hex": "#A0A0A0",
            "on": "#FFFFFF",
          },
          "tint80": Object {
            "RGB": "105,105,105",
            "hex": "#696969",
            "on": "#FFFFFF",
          },
        },
        "complimentary": Object {
          "dark": Object {
            "RGB": "206,185,0",
            "hex": "#CEB900",
            "on": "#272727",
          },
          "light": Object {
            "RGB": "255,245,157",
            "hex": "#FFF59D",
            "on": "#272727",
          },
          "main": Object {
            "RGB": "255,238,82",
            "hex": "#FFEE52",
            "on": "#272727",
          },
        },
        "error": Object {
          "dark": Object {
            "RGB": "197,73,1",
            "hex": "#C54901",
            "on": "#FFFFFF",
          },
          "main": Object {
            "RGB": "255,115,32",
            "hex": "#FF7320",
            "on": "#FFFFFF",
          },
        },
        "primary": Object {
          "dark": Object {
            "RGB": "102,22,20",
            "hex": "#661614",
            "on": "#FFFFFF",
          },
          "light": Object {
            "RGB": "218,52,47",
            "hex": "#DA342F",
            "on": "#FFFFFF",
          },
          "main": Object {
            "RGB": "164,40,36",
            "hex": "#A42824",
            "on": "#FFFFFF",
          },
        },
        "secondary": Object {
          "dark": Object {
            "RGB": "5,38,59",
            "hex": "#05263B",
            "on": "#FFFFFF",
          },
          "light": Object {
            "RGB": "89,156,175",
            "hex": "#599CAF",
            "on": "#FFFFFF",
          },
          "main": Object {
            "RGB": "11,70,107",
            "hex": "#0B466B",
            "on": "#FFFFFF",
          },
        },
        "surface": Object {
          "RGB": "255,255,255",
          "hex": "#FFFFFF",
          "on": "#272727",
        },
        "white": Object {
          "RGB": "255,255,255",
          "hex": "#FFFFFF",
          "on": "#272727",
        },
      }
    }
  >
    <span
      className="c0"
    />
  </StyledComponent>
</NoResults>
`;
